llvm/lib/Analysis/ConstantFolding.cpp

```diff
@@ -3826,6 +3827,37 @@ static Constant *ConstantFoldFixedVectorCall(
     }
     return ConstantVector::get(Result);
   }
+  case Intrinsic::wasm_dot: {
+    unsigned NumElements =
+        cast<FixedVectorType>(Operands[0]->getType())->getNumElements();
+
+    assert(NumElements == 8 && Result.size() == 4 &&
+           "wasm dot takes i16x8 and produces i32x4");
+    assert(Ty->isIntegerTy());
+    SmallVector<APInt, 8> MulVector;
```

Comment:
Nit, if we don't need the arbitrary precisioness of APInt or the dynamicness of a vector, we can just use a plain old int32_t array:


```suggestion
    int32_t MulVector[8];
```

And then you can use `ConstantInt::getSExtValue` to get the values out. It should be a little bit faster than using an APInt.

Reply:
very nice advice, ty Luke, will fix